src/scatter_buildblock/ScatterEstimation.cxx

```diff
@@ -0,0 +1,1486 @@
+
+    if (this->do_average_at_2)
+        act_image_for_averaging.reset(this->current_activity_image_sptr->clone());
+
+    //
+    // Begin the estimation process...
+    //
+    info("ScatterEstimation: Begin the estimation process...");
+    for (int i_scat_iter = 1;
+         i_scat_iter <= this->num_scatter_iterations;
+         i_scat_iter++)
+    {
+
+
+        if ( this->do_average_at_2)
+        {
+            if (i_scat_iter == 2) // do average 0 and 1
+            {
+                if (is_null_ptr(act_image_for_averaging))
+                    error("Storing the first activity estimate has failed at some point.");
+
+                *this->current_activity_image_sptr += *act_image_for_averaging;
+                *this->current_activity_image_sptr /= 2.f;
+            }
+        }
+
+        if (!iterative_method)
+        {
+            // Threshold
+        }
+
+        info("ScatterEstimation: Scatter simulation in progress...");
+
+        if (this->scatter_simulation_sptr->process_data() == Succeeded::no)
+            error("ScatterEstimation: Scatter simulation failed");
+        info("ScatterEstimation: Scatter simulation done...");
+
+        if(this->run_debug_mode) // Write unscaled scatter sinogram
+        {
+            std::stringstream convert;   // stream used for the conversion
+            convert << "unscaled_" << i_scat_iter;
+            FilePath tmp(convert.str(),false);
+            tmp.prepend_directory_name(extras_path.get_path());
+            unscaled_est_projdata_sptr->write_to_file(tmp.get_string());
+        }
+
+        // Set the min and max scale factors
+        if (i_scat_iter > 0)
+        {
+            local_max_scale_value = this->max_scale_value;
+            local_min_scale_value = this->min_scale_value;
+        }
+
+        scaled_est_projdata_sptr->fill(0.F);
+
+        upsample_and_fit_scatter_estimate(*scaled_est_projdata_sptr, *data_to_fit_projdata_sptr,
+                                          *unscaled_est_projdata_sptr,
+                                          *normalisation_factors_sptr,
+                                          *this->mask_projdata_sptr, local_min_scale_value,
+                                          local_max_scale_value, this->half_filter_width,
+                                          spline_type, true);
+
+        if(this->run_debug_mode)
+        {
+            std::stringstream convert;   // stream used for the conversion
+            convert << "scaled_" << i_scat_iter;
+            FilePath tmp(convert.str(),false);
+            tmp.prepend_directory_name(extras_path.get_path());
+            dynamic_cast<ProjDataInMemory *> (scaled_est_projdata_sptr.get())->write_to_file(tmp.get_string());
+        }
+
+        if (this->export_scatter_estimates_of_each_iteration ||
+                i_scat_iter == this->num_scatter_iterations -1 )
+        {
+
+            //this is complicated as the 2d scatter estimate was
```

Comment:
I set the `run_in_2d` check early in the `set_up()`.
There is no reason to reach here to get the error.

Reply:
actually this comment, in combination with the comment earlier that I do not update the additive data made me think, why you said that. 

What I aimed to do was to do 3D simulation and reconstruction and 2D `upsample_and_fit_scatter_estimate`. 

However, from your comments so far, I see that you want to run everything either in 2D or 3D.

Reply:
it doesn't seem to make any sense to do 3D simulation and 2D upsampling, but I agree at some point we want to be able to mix these things. However, ucrrently, `upsampling` will ignore any oblique sinograms in the the input data anyway (long standing limitation), so there is definitely no point right now. 

So, let's not make the code more complicated than the cases we actually want to cover at present, and just throw `error` otherwise. (no point implementing anything if we do/cannot test it)